presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RemoveRedundantCastToVarcharInJoinClause.java

```diff
@@ -0,0 +1,170 @@
+ * `bigint = try_cast(varchar as bigint)` as if the varchar here cannot be converted to bigint, try_cast will return null and will not match anyway. However, a special case is
+ * varchar begins with 0. `select cast(92 as varchar) = '092'` is false, but `select 92 = try_cast('092' as bigint)` returns true.
+ */
+public class RemoveRedundantCastToVarcharInJoinClause
+        implements Rule<JoinNode>
+{
+    private static final List<Type> TYPE_SUPPORTED = ImmutableList.of(INTEGER, BIGINT);
+    private final FunctionAndTypeManager functionAndTypeManager;
+    private final FunctionResolution functionResolution;
+
+    public RemoveRedundantCastToVarcharInJoinClause(FunctionAndTypeManager functionAndTypeManager)
+    {
+        this.functionAndTypeManager = functionAndTypeManager;
+        this.functionResolution = new FunctionResolution(functionAndTypeManager.getFunctionAndTypeResolver());
+    }
+
+    @Override
+    public boolean isEnabled(Session session)
+    {
+        return isRemoveRedundantCastToVarcharInJoinEnabled(session);
+    }
+
+    @Override
+    public Pattern<JoinNode> getPattern()
+    {
+        return join();
+    }
+
+    @Override
+    public Result apply(JoinNode node, Captures captures, Context context)
+    {
+        PlanNode leftInput = context.getLookup().resolve(node.getLeft());
+        PlanNode rightInput = context.getLookup().resolve(node.getRight());
+        if (!(leftInput instanceof ProjectNode) || !(rightInput instanceof ProjectNode)) {
+            return Result.empty();
+        }
+        ProjectNode leftProject = (ProjectNode) leftInput;
+        ProjectNode rightProject = (ProjectNode) rightInput;
+
+        ImmutableList.Builder<JoinNode.EquiJoinClause> joinClauseBuilder = ImmutableList.builder();
+        ImmutableMap.Builder<VariableReferenceExpression, RowExpression> newLeftAssignmentsBuilder = ImmutableMap.builder();
+        ImmutableMap.Builder<VariableReferenceExpression, RowExpression> newRightAssignmentsBuilder = ImmutableMap.builder();
+        boolean isChanged = false;
+        for (JoinNode.EquiJoinClause equiJoinClause : node.getCriteria()) {
+            RowExpression leftProjectAssignment = leftProject.getAssignments().getMap().get(equiJoinClause.getLeft());
+            RowExpression rightProjectAssignment = rightProject.getAssignments().getMap().get(equiJoinClause.getRight());
+            if (!isSupportedCast(leftProjectAssignment) || !isSupportedCast(rightProjectAssignment)) {
+                joinClauseBuilder.add(equiJoinClause);
+                continue;
+            }
+
+            RowExpression leftAssignment = ((CallExpression) leftProjectAssignment).getArguments().get(0);
+            RowExpression rightAssignment = ((CallExpression) rightProjectAssignment).getArguments().get(0);
+
+            if (!leftAssignment.getType().equals(rightAssignment.getType())) {
+                leftAssignment = castToBigInt(functionAndTypeManager, leftAssignment);
```

Comment:
we don't need to do this if the expressions are already bigints

Reply:
This is checked in `castToBigInt` function, with
```
        if (rowExpression.getType().equals(BIGINT)) {
            return rowExpression;
        }
```